content/en/docs/components/pipelines/sdk/python-function-components.ipynb

```diff
@@ -287,14 +287,55 @@
+    "\n",
+    "When you use the Kubeflow Pipelines SDK to convert your Python function to a\n",
+    "pipeline component, the Kubeflow Pipelines SDK uses the function's interface\n",
+    "to define the interface of your component in the following ways.\n",
+    "\n",
+    "*   Some arguments define input parameters.\n",
+    "*   Some arguments define output parameters.\n",
+    "*   The function's return value is used as an output parameter. If the return\n",
+    "    value is a [`collections.namedtuple`][named-tuple], the named tuple is used\n",
+    "    to return several small values. \n",
+    "\n",
+    "Since you can pass parameters between components as a value or as a path, the\n",
+    "Kubeflow Pipelines SDK removes common parameter suffixes that leak the\n",
+    "component's expected implementation. For example, a Python function-based\n",
+    "component that ingests data and outputs CSV data may have an output argument\n",
+    "that is defined as `csv_path: comp.OutputPath(str)`. In this case, the output\n",
+    "is the CSV data, not the path. So, the Kubeflow Pipelines SDK simplifies the\n",
+    "output name to `csv`.\n",
+    "\n",
+    "The Kubeflow Pipelines SDK uses the following rules to define the input and\n",
+    "output parameter names in your component's interface:\n",
+    "\n",
+    "*   If the argument name ends with `_path` and the argument is annotated as an\n",
+    "    [`kfp.components.InputPath`][input-path] or\n",
+    "    [`kfp.components.OutputPath`][output-path], the parameter name is the\n",
+    "    argument name with the trailing `_path` removed.\n",
+    "*   If the argument name ends with `_file`, the parameter name is the argument\n",
+    "    name with the trailing `_file` removed.\n",
+    "*   If you return a single small value from your component using the `return`\n",
+    "    statement, the output parameter is named `output`.\n",
+    "*   If you return several small values from your component by returning a \n",
+    "    [`collections.namedtuple`][named-tuple], the Kubeflow Pipelines SDK uses\n",
+    "    the tuple's field names as the output parameter names. \n",
+    "\n",
+    "Otherwise, the Kubeflow Pipelines SDK uses the argument name as the parameter\n",
+    "name.\n",
+    "\n",
     "<a name=\"pass-by-value\"></a>\n",
     "#### Passing parameters by value\n",
     "\n",
     "Python function-based components make it easier to pass parameters between\n",
     "components by value (such as numbers, booleans, and short strings), by letting\n",
     "you define your component’s interface by annotating your Python function. The\n",
-    "supported types are `int`, `float`, `bool`, and `string`. If you do not\n",
-    "annotate your function, these input parameters are passed as strings.\n",
+    "supported types are `int`, `float`, `bool`, and `str`. You can also pass \n",
+    "`list` or `dict` instances by value, if they contain small values, such as\n",
```

Comment:
>if they contain small values, such as `int`, `float`, `bool`, or `str` values.

Small correction: Lists and dictionaries can contain other lists and dictionaries. Basically they can contain anything that can be serialized/deserialized with `json.dumps`/`json.loads`.

Long version:
This is almost, but not 100% correct.
There are two core requirements:
* The total size of arguments consumed by value (inputValue) in the pipeline should not be big (<<1MiB per pipeline run).
* The constant arguments passed to a component needs to be serializable to string.
* The serialized string cannot contain NULL-bytes (`\x00`)

There are built-in serializers for the following types: `str` (`"String"`), `int` (`"Integer"`), `float` (`"Float"`), `bool`(`"Bool"`, `"Boolean"`), `list` (`List`, `"JsonArray"`), `dict` (`Dict`, `"JsonObject"`).

Here is what follows:
* Function parameters can be annotated with any python types or type names.
* By default, the function will receive the value of each parameter as a string.
* However if the function parameter is annotated with a known supported type (e.g. `int`, `float`, `bool`, `list`, `dict`), the parameter value will be automatically deserialized and the function will receive the value with a proper type (e.g. `int`).
* When returning the outputs by value, the returned object needs to be serialized to a string (if it's not already a string). Most values are serialized using the `str(value)` function. However for `list` and `dict`-typed outputs the `json.dumps(value)` serialization is used (if the output value is not a string already). `json.dumps` has certain limitations and quirks. It can only serialize a limited number of types. Also JSON considers all dictionary keys to be strings.

Some relevant code:
https://github.com/kubeflow/pipelines/blob/a0fdb92413c4dd97ff7be554b4df680907d2c56d/sdk/python/kfp/components/_data_passing.py#L108..L117